src/OutputWriters/checkpointer.jl

```diff
@@ -45,96 +43,105 @@ Keyword arguments
+function Checkpointer(model; frequency=nothing, interval=nothing, dir=".",
+                      prefix="checkpoint", force=false, verbose=false,
+                      properties = [:architecture, :boundary_conditions, :grid, :clock, :coriolis,
+                                    :buoyancy, :closure, :velocities, :tracers, :timestepper])
 
     validate_interval(frequency, interval)
 
-    # Grid needs to be checkpointed for restoring to work.
-    :grid ∉ properties && push!(properties, :grid)
-
-    has_array_refs = Symbol[]
+    # Certain properties are required for `restore_from_checkpoint` to work.
+    required_properties = (:grid, :architecture, :velocities, :tracers, :timestepper)
+    for rp in required_properties
+        if rp ∉ properties
+            @warn "$rp is required for checkpointing. It will be added to checkpointed properties"
+            push!(properties, rp)
+        end
+    end
 
     for p in properties
-        isa(p, Symbol) || @error "Property $p to be checkpointed must be a Symbol."
+        p isa Symbol || @error "Property $p to be checkpointed must be a Symbol."
         p ∉ propertynames(model) && @error "Cannot checkpoint $p, it is not a model property!"
 
         if has_reference(Function, getproperty(model, p))
             @warn "model.$p contains a function somewhere in its hierarchy and will not be checkpointed."
             filter!(e -> e != p, properties)
         end
-
-        has_reference(AbstractField, getproperty(model, p)) && push!(has_array_refs, p)
     end
 
     mkpath(dir)
-    return Checkpointer(frequency, interval, 0.0, dir, prefix, properties, has_array_refs, force, verbose)
+    return Checkpointer(frequency, interval, 0.0, dir, prefix, properties, force, verbose)
 end
 
 function write_output(model, c::Checkpointer)
-    filepath = joinpath(c.dir, c.prefix * string(model.clock.iteration) * ".jld2")
+    filepath = joinpath(c.dir, c.prefix * "_iteration" * string(model.clock.iteration) * ".jld2")
     c.verbose && @info "Checkpointing to file $filepath..."
 
-    t0 = time_ns()
+    t1 = time_ns()
     jldopen(filepath, "w") do file
         file["checkpointed_properties"] = c.properties
-        file["has_array_refs"] = c.has_array_refs
-
-        serializeproperties!(file, model, filter(e -> !(e ∈ c.has_array_refs), c.properties))
-        saveproperties!(file, model, filter(e -> e ∈ c.has_array_refs, c.properties))
+        serializeproperties!(file, model, c.properties)
     end
 
-    t1, sz = time_ns(), filesize(filepath)
-    c.verbose && @info "Checkpointing done: time=$(prettytime((t1-t0)/1e9)), size=$(pretty_filesize(sz))"
+    t2, sz = time_ns(), filesize(filepath)
+    c.verbose && @info "Checkpointing done: time=$(prettytime((t2-t1)/1e9)), size=$(pretty_filesize(sz))"
 end
 
+# This is the default name used in the simulation.output_writers ordered dict.
 defaultname(::Checkpointer, nelems) = :checkpointer
+
 _arr(::CPU, a) = a
```

Comment:
could call this function `convert_to_arch`?

Reply:
Done.